market/src/contract.rs

```diff
@@ -0,0 +1,169 @@
+use near_sdk::serde_json::json;
+use near_sdk::{env, log, near_bindgen};
+use near_sdk::{AccountId, Promise};
+use near_sdk::json_types::{Base64VecU8};
+use std::default::Default;
+
+use crate::consts::*;
+use crate::storage::*;
+
+impl Default for Market {
+    fn default() -> Self {
+        env::panic_str("Market should be initialized before usage")
+    }
+}
+
+#[near_bindgen]
+impl Market {
+    #[init]
+    pub fn new(market: MarketData, dao_account_id: AccountId) -> Self {
+        if env::state_exists() {
+            env::panic_str("ERR_ALREADY_INITIALIZED");
+        }
+
+        Self {
+            market,
+            dao_account_id,
+            resolved: false,
+            published: false,
+            closed: false,
+            proposals: Vec::new(),
+        }
+    }
+
+    pub fn get_market_data(&self) -> MarketData {
+        self.market.clone()
+    }
+
+    pub fn get_proposals(&self) -> Vec<u64> {
+        self.proposals.clone()
+    }
+
+    pub fn is_published(&self) -> bool {
+        self.published
+    }
+
+    pub fn is_resolved(&self) -> bool {
+        self.resolved
+    }
+
+    pub fn is_closed(&self) -> bool {
+        self.closed
+    }
+
+    pub fn is_market_expired(&self) -> bool {
+        self.market.expiration_date < env::block_timestamp().try_into().unwrap()
+    }
+
+    pub fn is_resolution_window_expired(&self) -> bool {
+        self.market.expiration_date + self.market.resolution_window < env::block_timestamp().try_into().unwrap()
+    }
+
+    #[payable]
+    pub fn publish_market(&mut self) -> Promise {
+        if self.published {
+            env::panic_str("ERR_MARKET_ALREADY_PUBLISHED");
+        }
+
+        if self.is_market_expired() {
+            env::panic_str("ERR_MARKET_EXPIRED");
+        }
+
+        //@TODO Research for an alternative to not create an empty Promise
+        let mut promises: Promise = Promise::new(self.dao_account_id.clone());
+        let mut count = 0;
+
+        for market_option in &self.market.options {
+            let args = Base64VecU8(json!({"response": count}).to_string().into_bytes());
```

Comment:
I'm not clear about this `count` arg to be passed into the `resolve` method? Is it like just a check of length?